---
title: An R Markdown document converted from "4_Genomic_Queries.ipynb"
output: html_document
---

# *NHLBI BioData CatalystÂ® (BDC)* Powered by PIC-SURE API Use-Case: Querying on Genomic Variables

This tutorial notebook focuses how to use the PIC-SURE API to query genomic and phenotypic data together.

For a more basic introduction to the python PIC-SURE API, see the `1_PICSURE_API_101.ipynb` notebook.
 
**Before running this notebook, please be sure to get a user-specific security token. For more information about how to proceed, see the \"Get your security token\" instructions in the [`README.md` file](../README.md).**

 -------

# Environment set-up

### Pre-requisites
- R 3.4 or later

### Install packages

**Note that if you are using the dedicated PIC-SURE environment within the *BDC Powered by Seven Bridges (BDC-Seven Bridges)* platform, the necessary packages have already been installed.**

*Troubleshooting: if you encounter an error while installing packages or connecting to a PIC-SURE resource, try restarting your kernel and trying again.*

```{r}
### Uncomment this code if you are not using the PIC-SURE environment in *BDC-Seven Bridges*, or if you do not have all the necessary dependencies installed.
install.packages(c("devtools", "dplyr"))
```

```{r}
#Sys.setenv(TAR = "/bin/tar")
#options(unzip = "internal")
devtools::install_github("hms-dbmi/pic-sure-r-adapter-hpds", ref="main", force=T, quiet=FALSE)
library(dplyr)
library(picsure)
```

### Connecting to a PIC-SURE resource

```{r}
PICSURE_network_URL = "https://picsure.biodatacatalyst.nhlbi.nih.gov/picsure"
token_file <- "token.txt"
token <- scan(token_file, what = "character")
session <- picsure::bdc.initializeSession(PICSURE_network_URL, token)
session <- picsure::bdc.setResource(session = session,  resourceName = "AUTH")
     
```

# Example 1: Build a query with genomic and phenotypic variables
## 1.1 Building the query with the PIC-SURE API

We are going to create a new query request from the PIC-SURE resource that was specified above.  For this example, we will limit the query to a single dataset (the TOPMed DCC Harmonized dataset) by filtering on sex, body mass index, and two genomic filters. 

For more information about the TOPMed DCC Harmonized dataset, please refer to the [`2_TOPMed_DCC_Harmonized_Variables_analysis.ipynb` notebook](./2_TOPMed_DCC_Harmonized_Variables_analysis.ipynb).

First we will create a new query instance.

```{r}
my_query <- picsure::bdc.newQuery(session) # Initialize a new query
```

### Find all TOPMed DCC Harmonized variables
We can search for variables related to our search query using the `bdc::find.in.dictionary` function. 

In this example, we will retrieve all variables available in the TOPMed DCC Harmonized dataset.

You can find information about the phs number associated with each study and what data are available from the Data Access Dashboard in the PIC-SURE [User Interface](https://picsure.biodatacatalyst.nhlbi.nih.gov/).

```{r}
harmonized_df <- picsure::bdc.searchPicsure(session, 'DCC Harmonized', includeValues = TRUE)  
```

## 1.2 Add categorical phenotypic variable (sex) to the query

First, we will search our TOPMed DCC Harmonized data dictionary for a sex-related variable.

```{r}
sex_var <- harmonized_df %>% filter(grepl('subject sex', var_description, ignore.case = TRUE))
sex_var
```

Let's examine the values within this variable.

```{r}
sex_var %>% pull(values)
```

We are only interested in Female values in this use case - let's apply this filter to our query accordingly.

```{r}
my_query <- picsure::addClause(query = my_query, 
                               keys = sex_var %>% pull(name),
                               type = 'FILTER',
                               categories = list('Female'))
```

## 1.3 Add continuous phenotypic variable (body mass index, BMI) to the query
For this example, we are only interested in obese participants (BMI >= 30).

Following the data dictionary search pattern shown above, we can search for the TOPMed DCC Harmonized dataset variables related to BMI.

```{r}
# Search our TOPMed harmonized dataframe for an age variable
bmi_var <- harmonized_df %>% filter(grepl('body mass index', var_description, ignore.case = TRUE))
print(paste0(bmi_var$var_name, ': ', bmi_var$name))

# Examine the values
print(paste0('Min: ', bmi_var$min, ', Max: ', bmi_var$max))

# Filter to obese participants with BMI 30 or more
my_query <- picsure::addClause(query = my_query, 
                               keys = bmi_var$name,
                               type = 'FILTER',
                               min = 30)
```

## 1.4 Add genomic filters to the query
To start adding genomic filters to our query, we first need to understand what genomic annotations are available.

```{r}
genotype_annotations <- picsure::getGenomicAnnotations(session)
genotype_annotations
```

As shown in the genomic_annotations object above, some genomic variables that can be used in queries include `Gene_with_variant`, `Variant_class`, and `Variant_severity`.

#### Add genotypic variable (Gene_with_variant) to the query

By looking at the values associated with the `Gene_with_variant` annotation we see what genes we can filter the variants by.

The fat mass and obesity related gene, or *FTO* gene, [has been linked to obesity and other diseases](https://www.ncbi.nlm.nih.gov/gene/79068). Let's search for this gene.

```{r}
# get total list of genes
genes <- genotype_annotations %>% filter(name == 'Gene_with_variant') %>% pull(values)
genes <- strsplit(genes, ', ')[[1]]

# print the first 10 genes
print(genes[1:10])

# check if a certain gene of interest, e.g. FTO, is in the gene list
gene_of_interest <- 'FTO'
gene_of_interest %in% genes
```

Let's also say that we only want highly severe variants. We can see the levels of severity shown below.

```{r}
# what values are available for Variant_severity?
genotype_annotations %>% filter(name == "Variant_severity") %>% pull(values)
```

The gene list shown above provides a list of values that can be used for the `Gene_with_variant`, in this case genes affected by a variant. Let's say we are interested in participants who have a highly severe *FTO* variant. We can add these filters to the query.

```{r}
my_query <- picsure::addClause(query = my_query, 
                               keys = 'Gene_with_variant',
                               type = 'FILTER',
                               categories = list('FTO'))

my_query <- picsure::addClause(query = my_query,
                               keys = 'Variant_severity',
                               type = 'FILTER',
                               categories = list("HIGH"))
```


## 1.5 Retrieve data from the query

#### Getting query count
We have now built a query called `my_query` which contains the search criteria we are interested in:
- Sex = Female
- BMI > 30
- Participants have a highly severe *FTO* variant

We will run a count query to find the number of matching participants.
This is a great way to preview how many participants match your query criteria without extracting the data.

```{r}
picsure::runQuery(my_query, resultType = 'Count')
```

#### Getting query data
We will retrieve our results in a dataframe.
Note that since we only added sex and age to the query, these are the only phenotypic variables returned.

```{r}
query_result <- picsure::runQuery(my_query)
```

```{r}
dim(query_result)
```

```{r}
head(query_result)
```

# Example 2: Use case with *SERPINA1* gene variants as a risk factor for  COPD

In this example, we will create a query to explore the relationship between the COPD phenotype and variants in the SERPINA1 gene. Variations of the SERPINA1 gene have been found to be a strong risk factor for COPD, which you can read more about [here](https://pubmed.ncbi.nlm.nih.gov/31661293/).

To explore this relationship, we will narrow the cohort down to participants that meet the following criteria:
* participated in the COPDgene study
* have had COPD
* have a missense variant of the *SERPINA1* gene

## 2.1 Create query
Let's start by creating a new query and finding the variables pertaining to the COPDgene study (phs000179)

```{r}
# Retrieve all COPDGene variables
copd_df <- picsure::bdc.searchPicsure(session, 'phs000179', includeValues = TRUE)  
```

### Criteria 1: Participants who have had COPD
Let's search our dataframe of COPDGene variables to find the one we want to use to filter for participants who have ever had COPD.

```{r}
copd_df %>% 
    filter(grepl('copd', var_description, ignore.case = TRUE)) %>%
    select(name, var_name, var_description, values)
```

The variable with the description "COPD: have you ever had COPD" seems the most promising. Let's copy that variable's HPDS path to use in our query. Also note what values are present for this variable- we will filter to participants with value "Yes".

```{r}
copd_path <- '\\phs000179\\pht002239\\phv00159731\\COPD\\'
```

```{r}
copd_query <- picsure::bdc.newQuery(session) # Initialize a new query

copd_query <- picsure::addClause(query = copd_query, 
                                 keys = copd_path,
                                 type = 'FILTER',
                                 categories = list('Yes'))
```

### Criteria 2: Participants with a missense variant on *SERPINA1*

A filter for the *SERPINA1* gene can be added in a way similar to the first query shown above.

```{r}
copd_query <- picsure::addClause(query = copd_query, 
                                 keys = 'Gene_with_variant',
                                 type = 'FILTER',
                                 categories = list('SERPINA1'))
```

We can take a look at the variant consequences available for filtering to confirm that missense variant is a filtering option.

```{r}
# get all "Variant_consequence_calculated" values
consequences = genotype_annotations %>% filter(name == 'Variant_consequence_calculated') %>% pull(values)
consequences <- strsplit(consequences, ', ')[[1]]
print(consequences)
```

```{r}
copd_query <- picsure::addClause(query = copd_query, 
                                 keys = 'Variant_consequence_calculated',
                                 type = 'FILTER',
                                 categories = list("missense_variant"))
```

### 2.2 Get Results
Now that the filtering is complete, we can use this final query to get counts and perform analysis on the data.

```{r}
picsure::runQuery(copd_query, resultType = 'count')
```

```{r}
copd_result <- picsure::runQuery(copd_query)
```

```{r}
dim(copd_result)
```

```{r}
head(copd_result)
```

